src/data/DataStore.ts

```diff
@@ -983,6 +985,10 @@ class DataStore {
                 }
             }
 
+            if (appendNaNList.length && appendNaNList.length !== frameEnd - frameStart) {
+                newIndices[sampledIndex++] = appendNaNList[0]; // append first NaN point in every bucket
```

Comment:
If we only append the first NaN point. We should use a variable instead of an array because array will cost more. For example:

```ts
let firstNaNIndex = -1;

if (isNaN(y) && firstNaNIndex < 0) {
  firstNaNIndex = rawIndex.
)
...
if (firstNaNIndex >= 0) {
  newIndices[sampledIndex++] = firstNaNIndex;
}
```

The performance is one of the most important indicator when implementing the sampling algorithm. We can check if it's degenerated in the `test/sample-compare.html` case

Reply:
yes, got it.